### ISSUE TITLE
Vulnerabilities for Python Upstream dependencies

### ISSUE THREAD

**Issue by krishnapaparaju** [quay/clair#591]
**Description of Problem**:

Clair needs integration with a vulnerability database to display vulnerabilities for Python  upstream dependencies. 

**Expected Outcome**:

Integrate with OpenShift.IO AI platform as another 'Vulnerability Metadata source' to track upstream vulnerability metadata for Python dependencies. CVE IDs are mapped from NVD to specific Python dependencies, this work is tracked [here](https://github.com/fabric8-analytics/cvejob).

1) Get a OpenShift.IO API token under 'profile' section   (Require registration at [OpenShift.IO](openshift.io))
2) Find any vulnerabilities for a Python application dependency
      _curl -k -H "Authorization: Bearer $token"  https://recommender.api.openshift.io/api/v1/component-analyses/pypi/dependency-name/version_
      (For Python dependency pycrypto:2.6.1, check out [API response](https://gist.github.com/krishnapaparaju/ea736e2f71f12f0582e18e9d34042395) showing associated vulnerabilities)

**Actual Outcome**:
Currently scanning does not cover vulnerabilities for Python upstream dependencies.

**Environment**:
NA

**Comment by KeyboardNerd** [comment]
Since there are two different issues with nearly the same content, I'll comment only once: 
Link to another issue: #590 

1. It looks interesting but please point me to the API/usage documentation of the `https://recommender.api.openshift.io/api/v1/component-analyses/` end point to see if it's compatible with Clair. 
2. The language wide package vulnerability scanning may be added in the future but currently it's not supported.
3. What is the upstream vulnerability data source in this end point and how can we ensure that those vulnerabilities are related correctly?

Thanks.

**Comment by spaparaju** [comment]
1. Here is the swagger spec. for this end point:  https://raw.githubusercontent.com/fabric8-analytics/fabric8-analytics-server/master/swagger.yaml

3. Here are the docs. covering the process for mapping CVEs to specific application dependencies and roadmap to onboard more ecosystems: https://github.com/fabric8-analytics/cvejob/tree/master/docs

**Comment by samuzzal-choudhury** [comment]
Here is a link to try out component-analyses using the swagger spec:

http://analytics-bayesian-coreapi.dev.rdu2c.fabric8.io/api/v1/ui/#!/Scan_Services/f8a_scanner_api_v1_component_analyses

This communicates with the production deployed analytics service to fetch component data. One requires to provide one's openshift.io user token in the spec to authorize the call.

**Comment by spaparaju** [comment]
cc @KeyboardNerd

**Comment by jzelinskie** [comment]
Quoting myself from the other thread:

>So far, Clair has only supported vulnerability sources from official sources for the given namespace. For example, Node should probably use the same API as npm audit. I'm also hesitant to add any vulnerability sources gated behind API keys.
>
>I think this is a reasonable thing to implement, but I'm not sure it makes sense to upstream.

**Comment by spaparaju** [comment]
Mapping process CVEs to specific application dependencies varies a lot for each programming language. Couple of points to note:

- 'npm audit' audits a given application which different from the pull model of Clair (periodically pull in all the new CVEs from a source to PG)
-   Mapping a CVE to an application dependency is very different for each language (eg: For GO language, there is hardly any coverage at NVD, OpenShift.IO team is getting started with NLP on Git Commit messages) 
- Currently there no official sources (opensource, no license terms) for covering CVEs for application level dependencies. OpenShift.IO aims to cover CVEs for application dependencies spanning multiple languages (including integrations with 'npm audit') thru one OpenShift.IO API endpoint.
- API keys are hindrance to be viable on upstream, team is currently removing the need for API keys to reach OpenShift.IO platform.
- Team is figuring out the data model and components of Clair to start working here, upstream.

For all the ongoing activities, soon I will be sharing the relevant issues, documentation.

**Comment by spaparaju** [comment]
@jzelinskie couple of updates..

- API token restriction been removed, check out the  [Swagger spec](https://app.swaggerhub.com/apis/slenka/OSIO_API/0.1) (under CVE services)  
- A new Clair driver for CVEs related to NPM  been proposed, check out the [PR](https://github.com/coreos/clair/pull/676)
- As you suggested, this integration might not be viable for upstream Clair (license terms for CVEs from the proposed NPM Clair driver)

**Comment by jsfan** [comment]
While this issue only mentions Python dependencies, the discussion seem to have been revolving around Python and Node.js dependencies.

If it was only about Python, another option would be https://github.com/pyupio/safety-db which comes with a JSON file that could be used straight off GitHub. The file is licensed under CC BY-NC-SA 4.0. Would that be an acceptable license for data used by Clair?
